politeiawww/events.go

```diff
@@ -0,0 +1,404 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/google/uuid"
+
+	"github.com/decred/politeia/politeiawww/api/v1"
+	"github.com/decred/politeia/politeiawww/database"
+)
+
+type EventT int
+
+type EventManager struct {
+	Listeners map[EventT][]chan interface{}
+}
+
+const (
+	EventTypeInvalid                EventT = 0
+	EventTypeProposalSubmitted      EventT = 1
+	EventTypeProposalStatusChange   EventT = 2
+	EventTypeProposalEdited         EventT = 3
+	EventTypeProposalVoteStarted    EventT = 4
+	EventTypeProposalVoteAuthorized EventT = 5
+	EventTypeProposalVoteFinished   EventT = 6
+	EventTypeUserManage             EventT = 7
+)
+
+type EventDataProposalSubmitted struct {
+	CensorshipRecord *v1.CensorshipRecord
+	ProposalName     string
+	User             *database.User
+}
+
+type EventDataProposalStatusChange struct {
+	Proposal          *v1.ProposalRecord
+	SetProposalStatus *v1.SetProposalStatus
+	AdminUser         *database.User
+}
+
+type EventDataProposalEdited struct {
+	Proposal *v1.ProposalRecord
+}
+
+type EventDataProposalVoteStarted struct {
+	AdminUser *database.User
+	StartVote *v1.StartVote
+}
+
+type EventDataProposalVoteAuthorized struct {
+	AuthorizeVote *v1.AuthorizeVote
+	User          *database.User
+}
+
+type EventDataUserManage struct {
+	AdminUser  *database.User
+	User       *database.User
+	ManageUser *v1.ManageUser
+}
+
+func (b *backend) getProposalAuthor(proposal *v1.ProposalRecord) (*database.User, error) {
+	if proposal.UserId == "" {
+		proposal.UserId = b.userPubkeys[proposal.PublicKey]
+	}
+
+	userID, err := uuid.Parse(proposal.UserId)
+	if err != nil {
+		return nil, fmt.Errorf("parse UUID: %v", err)
+	}
+	return b.db.UserGetById(userID)
+}
+
+func (b *backend) initEventManager() {
+	b.eventManager = &EventManager{}
+
+	b.setupProposalSubmittedEmailNotification()
+
+	b.setupProposalStatusChangeEmailNotification()
+	b.setupProposalStatusChangeLogging()
+
+	b.setupProposalEditedEmailNotification()
+
+	b.setupProposalVoteStartedEmailNotification()
+	b.setupProposalVoteStartedLogging()
+
+	b.setupProposalVoteAuthorizedEmailNotification()
+
+	b.setupUserManageLogging()
+}
+
+func (b *backend) setupProposalSubmittedEmailNotification() {
+	if b.cfg.SMTP == nil {
+		return
+	}
+
+	ch := make(chan interface{})
+	go func() {
+		for {
```

Comment:
The typical way to do this is just to range on the channel.

```Go
for data := range ch {
...
}
```

Same deal with all the others.